src/functional/otf/compilation/compiler.py

```diff
@@ -0,0 +1,91 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
+
+import dataclasses
+import pathlib
+from typing import Protocol, TypeVar
+
+from functional.otf import languages, stages, step_types, workflow
+from functional.otf.compilation import build_data, cache, importer
+from functional.otf.step_types import LS, SrcL, TgtL
+
+
+SourceLanguageType = TypeVar("SourceLanguageType", bound=languages.LanguageTag)
+LanguageSettingsType = TypeVar("LanguageSettingsType", bound=languages.LanguageSettings)
+T = TypeVar("T")
+
+
+def is_compiled(data: build_data.BuildData) -> bool:
+    return data.status >= build_data.BuildStatus.COMPILED
+
+
+def module_exists(data: build_data.BuildData, src_dir: pathlib.Path) -> bool:
+    return (src_dir / data.module).exists()
+
+
+class BuildSystemProjectGenerator(Protocol[SrcL, LS, TgtL]):
+    def __call__(
+        self,
+        source: stages.CompilableSource[SrcL, LS, TgtL],
+        cache_strategy: cache.Strategy,
+    ) -> stages.BuildSystemProject[SrcL, LS, TgtL]:
+        ...
+
+
+@dataclasses.dataclass(frozen=True)
+class Compiler(
+    step_types.CompilationStep[SourceLanguageType, LanguageSettingsType, languages.Python]
+):
+    cache_strategy: cache.Strategy
+    builder_factory: BuildSystemProjectGenerator[
+        SourceLanguageType, LanguageSettingsType, languages.Python
+    ]
+    force_recompile: bool = False
+    """Use any build system (via configured factory) to compile a GT4Py program to a ``functional.otf.stages.CompiledProgram``."""
+
+    def __call__(
+        self,
+        inp: stages.CompilableSource[SourceLanguageType, LanguageSettingsType, languages.Python],
+    ) -> stages.CompiledProgram:
+        src_dir = cache.get_cache_folder(inp, self.cache_strategy)
+
+        data = build_data.read_data(src_dir)
+
+        if not data or not is_compiled(data) or self.force_recompile:
+            self.builder_factory(inp, self.cache_strategy).build()
+
+        new_data = build_data.read_data(src_dir)
+
+        if not new_data or not is_compiled(new_data) or not module_exists(new_data, src_dir):
+            raise CompilerError(
+                "On-the-fly compilation unsuccessful for {inp.source_module.entry_point.name}!"
+            )
```

Comment:
Can you include the STDOUT & STDERR from the build system?

Reply:
You mean trying to read the logs and handling all possible cases of logs not being written / corrupted in the error handling?

I would prefer a message on how to proceed: set the cache path manually, run again and then look at the logs.

Reply:
No-no, I used to include the `cmake` subprocess' standard output and standard error here. You could do it similarly with the compile DB, only you'd include `gcc`'s std out.

Reply:
The changes required would be:
1) change the build system interface for `.build()` to return CLI output.
2) change build system implementations to gather, log and return stdout and stderr from all commands.
3) research how to helpfully convey / truncate the stdout and stderr in the error message. This is necessary because these might well overflow the buffer in some terminals (depending on build system flags, compilers available etc).

The simple logging approach was chosen to avoid all of that, while achieving almost the same amount of debuggability, if not the same. If you run in a debugger the cache folder will anyway live until you terminate the session.

Reply:
If you want I can put the above either in the ADR or in an issue.

Reply:
I find it quite handy to see the compiler messages in the unit test output. Sometimes it saves me going for the debugger and I can fix the issue right away. Can you perhaps make an issue for the general improvement of native code compilation error messages? It can be open-ended and we can discuss the approach to take outside of this PR.

Reply:
sure

Reply:
Issue is here https://github.com/GridTools/gt4py/issues/987